server/src/main/java/org/opensearch/plugins/LabelingPlugin.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.action.IndicesRequest;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.querygroup.LabelingService.LabelingImplementationType;
+
+/**
+ * This plugin introduces contracts on how should the incoming requests be labeled using implicit/explicit request attributes
+ *
+ */
+public interface LabelingPlugin {
+
+    LabelingImplementationType getImplementationName();
+
+    /**
+     * This method will compute  label value/values and
+     * put these in the {@link ThreadContext} against {@link org.opensearch.querygroup.LabelingHeader}
+     * @param request
+     * @param threadContext
+     */
+    void labelRequest(final IndicesRequest request, final ThreadContext threadContext);
```

Comment:
Are we planning to create a plugin for the query grouping feature? In which case it would be like?
```java
public class QueryInsightsPlugin extends Plugin implements LabelingPlugin 
{
...
}
```

Reply:
I think this doesn't concern the specific feature. Yes, This feature is supposed to be plugin driven to maintain segregation and strong cohesion with the features.

Reply:
Have you considered trying to use the [ActionPlugin.getActionFilters](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L94-L99) extension point? This allows you to run a filter before a transport action is executed on a node.

Here's an example of how Security [overrides it](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ActionPlugin.java#L94-L99) and registers a SecurityFilter. The filter needs to implement a method called [apply](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java#L155-L167) that has access to the Request object. You can also instantiate a filter within `createComponents` where you have access to the `threadPool` that core passes to plugins.

Reply:
💯 this is the way

Reply:
I see, I missed this comment. 
But the question I have is that is it okay to leave the `ThreadContext` open to plugins ?

Reply:
> But the question I have is that is it okay to leave the ThreadContext open to plugins ?

No and it is not necessary I believe

Reply:
I see the following code to sort the filters based on the `order()` method implementation
```
 public ActionFilters(Set<ActionFilter> actionFilters) {
        this.filters = actionFilters.toArray(new ActionFilter[0]);
        Arrays.sort(filters, new Comparator<ActionFilter>() {
            @Override
            public int compare(ActionFilter o1, ActionFilter o2) {
                return Integer.compare(o1.order(), o2.order());
            }
        });
    }
```

Reply:
> Security must be the first action filter to run since that is where authz is performed and that is also where the authenticated user info is serialized to the thread context

Security uses `RequestHandlerWrapper` if I am not mistaken

Reply:
@reta Authentication is done in the RestHandlerWrapper and authorization is done as an action filter.

Since 2.11, authentication has been moved further up the netty pipeline into the [header_verifier](https://github.com/opensearch-project/OpenSearch/blob/501a7024d18f7c7bc135b5e686835f0bcdf00d30/modules/transport-netty4/src/main/java/org/opensearch/http/netty4/Netty4HttpServerTransport.java#L463) step.

Reply:
@kaushalmahi12 Security reserves the lowest order (first)

```
@Override
public int order() {
    return Integer.MIN_VALUE;
}
```

Any other filter should have a higher value to occur after the [SecurityFilter](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java)

The SecurityFilter calls on [PrivilegesEvaluator.evaluate](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java#L381) which is the method in which the [authenticated user is serialized into the ThrreadContext](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L215-L227)

Reply:
> @reta Authentication is done in the RestHandlerWrapper and authorization is done as an action filter.

Ah I see, thanks @cwperks , I think the authentication does set the principal, right?